### ISSUE TITLE
Build error on Ubuntu

### ISSUE THREAD

**Issue by ThatcherC** [wwmm/easyeffects#1000]
Just tried building v6.0.0 on Ubuntu and encountered this vague error. Is this a known issue? Happy to try out any proposed fixes and report back. Thanks!

```
$ git checkout v6.0.0 
$ debuild -i -us -uc
 dpkg-buildpackage -us -uc -ui -i
dpkg-buildpackage: info: source package pulseeffects
dpkg-buildpackage: info: source version 4.8.2-1
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Mikhail Novosyolov <mikhailnov@dumalogiya.ru>
 dpkg-source -i --before-build .
dpkg-buildpackage: info: host architecture amd64
dpkg-source: error: source package has two conflicting values - easyeffects and pulseeffects
dpkg-buildpackage: error: dpkg-source -i --before-build . subprocess returned exit status 25
debuild: fatal error at line 1182:
dpkg-buildpackage -us -uc -ui -i failed
```

**Platform:**
- Ubuntu 20.10
- Linux 5.8.0-59-generic

**Comment by wwmm** [comment]
Our scripts for Debian packages still have to be updated. As I do not use a Debian based distribution I can not fix that.

**Comment by ThatcherC** [comment]
Okay! Thanks for the speedy response. I'll poke around and see if I can get a working build but I'm not an expert on build systems. Out of curiosity, what distributions are you developing on? And are there any older releases that would be better for a recent Debian? It might be nice to have a list of supported platforms/distros in the README. Really hoping I can this up and running!

**Comment by wwmm** [comment]
>what distributions are you developing on?

I run Arch Linux on all my computers.

>And are there any older releases that would be better for a recent Debian?

It depends. If you are using Pulseaudio it would be `4.8.5`. If you are using PipeWire and you can not build `6.0.0` then the next option is `5.0.4`

**Comment by jinwk00** [comment]
> It depends. If you are using Pulseaudio it would be `4.8.5`. If you are using PipeWire and you can not build `6.0.0` then the next option is `5.0.4`

Apparently the script for Debian has not been updated since 4.8.2-1
5.0.4:
` dpkg-buildpackage -us -uc -ui -i
dpkg-buildpackage: info: source package pulseeffects
dpkg-buildpackage: info: source version 4.8.2-1
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Mikhail Novosyolov <mikhailnov@dumalogiya.ru>
 dpkg-source -i --before-build .
dpkg-buildpackage: info: host architecture amd64
 fakeroot debian/rules clean
dh clean --with scour
   dh_clean
	rm -f debian/debhelper-build-stamp
	rm -rf debian/.debhelper/
	rm -f -- debian/pulseeffects.substvars debian/gstreamer1.0-convolver-pulseeffects.substvars debian/gstreamer1.0-crystalizer-pulseeffects.substvars debian/gstreamer1.0-autogain-pulseeffects.substvars debian/gstreamer1.0-adapter-pulseeffects.substvars debian/files
	rm -fr -- debian/pulseeffects/ debian/tmp/ debian/gstreamer1.0-convolver-pulseeffects/ debian/gstreamer1.0-crystalizer-pulseeffects/ debian/gstreamer1.0-autogain-pulseeffects/ debian/gstreamer1.0-adapter-pulseeffects/
	find .  \( \( \
		\( -path .\*/.git -o -path .\*/.svn -o -path .\*/.bzr -o -path .\*/.hg -o -path .\*/CVS -o -path .\*/.pc -o -path .\*/_darcs \) -prune -o -type f -a \
	        \( -name '#*#' -o -name '.*~' -o -name '*~' -o -name DEADJOE \
		 -o -name '*.orig' -o -name '*.rej' -o -name '*.bak' \
		 -o -name '.*.orig' -o -name .*.rej -o -name '.SUMS' \
		 -o -name TAGS -o \( -path '*/.deps/*' -a -name '*.P' \) \
		\) -exec rm -f {} + \) -o \
		\( -type d -a -name autom4te.cache -prune -exec rm -rf {} + \) \)
 dpkg-source -i -b .
dpkg-source: info: using source format '3.0 (git)'
dpkg-source: error: uncommitted, not-ignored changes in working directory: pulseeffects-build-deps_4.8.2-1_amd64.buildinfo pulseeffects-build-deps_4.8.2-1_amd64.changes
dpkg-buildpackage: error: dpkg-source -i -b . subprocess returned exit status 255
debuild: fatal error at line 1182:
dpkg-buildpackage -us -uc -ui -i failed
`

5.0.3:
` dpkg-buildpackage -us -uc -ui -i
dpkg-buildpackage: info: source package pulseeffects
dpkg-buildpackage: info: source version 4.8.2-1
dpkg-buildpackage: info: source distribution unstable
dpkg-buildpackage: info: source changed by Mikhail Novosyolov <mikhailnov@dumalogiya.ru>
 dpkg-source -i --before-build .
dpkg-buildpackage: info: host architecture amd64
 fakeroot debian/rules clean
dh clean --with scour
   dh_clean
	rm -f debian/debhelper-build-stamp
	rm -rf debian/.debhelper/
	rm -f -- debian/pulseeffects.substvars debian/gstreamer1.0-convolver-pulseeffects.substvars debian/gstreamer1.0-crystalizer-pulseeffects.substvars debian/gstreamer1.0-autogain-pulseeffects.substvars debian/gstreamer1.0-adapter-pulseeffects.substvars debian/files
	rm -fr -- debian/pulseeffects/ debian/tmp/ debian/gstreamer1.0-convolver-pulseeffects/ debian/gstreamer1.0-crystalizer-pulseeffects/ debian/gstreamer1.0-autogain-pulseeffects/ debian/gstreamer1.0-adapter-pulseeffects/
	find .  \( \( \
		\( -path .\*/.git -o -path .\*/.svn -o -path .\*/.bzr -o -path .\*/.hg -o -path .\*/CVS -o -path .\*/.pc -o -path .\*/_darcs \) -prune -o -type f -a \
	        \( -name '#*#' -o -name '.*~' -o -name '*~' -o -name DEADJOE \
		 -o -name '*.orig' -o -name '*.rej' -o -name '*.bak' \
		 -o -name '.*.orig' -o -name .*.rej -o -name '.SUMS' \
		 -o -name TAGS -o \( -path '*/.deps/*' -a -name '*.P' \) \
		\) -exec rm -f {} + \) -o \
		\( -type d -a -name autom4te.cache -prune -exec rm -rf {} + \) \)
 dpkg-source -i -b .
dpkg-source: error: can't build with source format '3.0 (git)': doesn't contain a git repository
dpkg-buildpackage: error: dpkg-source -i -b . subprocess returned exit status 25
debuild: fatal error at line 1182:
dpkg-buildpackage -us -uc -ui -i failed
`

**Comment by wunderbar78** [comment]
how about updating the build from source instructions and dependencies (before releasing an update)?! Even the linked Arch PKGBUILD is outdated!

horrible documentation!

**Comment by wwmm** [comment]
>how about updating the build from source instructions and dependencies (before releasing an update)?! 

Because I am not your slave and I do not remember seeing you helping me when I spent almost 5 months working alone and for free in the new release. Go ahead and update the build instructions and the documentation.

**Comment by wunderbar78** [comment]
if I stumble across this project 3 days ago, how should I help you in the last 5 months?
and no, you're not my slave, you're just horrible at documentation (or you just don't give af) if you can't even write down the dependencies. You have written all the stuff, took you 5 months, you know what your software depends on and I bet it takes you minutes to write them down. And now I should waste hours (or better 5 months?) and guess around what every error wants to tell me? You think I'm stupid? (salty cause of the slave bs) 

I'm not a software developer who can debug every error. 
I just would like to know the current dependencies and you can't find them anywhere around.
The README.md says that the dependencies changed, yet every documentation is for the old versions with outdated dependencies. 
I had a look at the linked Arch PKGBUILD (the distribution you are using) to get the dependencies, guess what, didn't help me cause outdated. 
that's why I ended here writing a salty comment, cause I already wasted time cause of missing documentation and missing knowledge how to troubleshoot all the stuff. yet you alone have written the software, know everything about it and are too arrogant to share them? Or what is it? Do I have to be an elitist Arch user to use this software?

funny thing is, I'm normally chill af, but you're driving me nuts, mate.
I would just like to use the software like others in this thread...

**Comment by partofthething** [comment]
I explored this on a Ubuntu 21.04 system and made some progress but didn't quite finish. 

* In the deb folder, rename all files with the word `pulseeffects` in them to `easyeffects`
```
mv pulseeffects.1 easyeffects.1
mv pulseeffects.install easyeffects.install
mv pulseeffects.manpages easyeffects.manpages
```
(note that some of these are git ignored so I had to add them with `git add -f` in my branch)

* Upgrade meson: `pip3 install --upgrade --user meson`
* In the deb folder changelog, find/replace pulseeffects with easyeffects (not sure if necessary)

This made it past the above error but then failed because the system repo had installed an older version of  pipewire that wasn't supported by the latest release here, v6.0.1. At that point I gave up, checked out the pre-renamed `v5.0.1`, and built without trouble.

**Comment by chrisabbeytx** [comment]
Some people spend hours putting on makeup just to stay home and look at themselves in the mirror.  Not everyone who creates a thing cares at all if anyone else gets to use it or enjoys it.  Fair enough.  If I want a parametric EQ program, I can learn how to code and write one.

**Comment by wwmm** [comment]
>had installed an older version of pipewire that wasn't supported by the latest release here, v6.0.1

Support for filter latency reporting was added to PipeWire recently. So that was unavoidable. 

>It will be easier if the debian official repository will updated its own releases:

Maybe. But I have my doubts it would be by much. In order to implement what I wanted the way I wanted I had to use recent versions of some important libraries. PipeWire can not be old. As I am using c++20 features gcc and Meson cannot be old. And the distribution must have gtkmm4.  

Based on what I have seen so far the hardest to get in most distributions is gtkmm4. As there aren't many "big programs" using gtkmm the distributions do not seem in a hurry to offer gtkmm4 packages. Even on Arch Linux it took months after the gtkmm4 release for it to be available in the stable repository. While at the same time gtk4 got there a lot faster because of gnome 40 would use it.

I also wonder if I should keep Debian build scripts here in our sources. As I do not use a Debian based distribution I can not maintain these scripts. I do not even know how Debian package creation works. The only reason why these scripts are here is because it seems that this make it easier for Debian users to create a package. What does not bother me. But the side effect is that people somehow expect me to help when these script do not work. And while I would like to help I cannot because I use Arch Linux.

**Comment by hosiet** [comment]
Renaming and migration to PipeWire/GTK4 is a great chance of renovation. As the packager in official Debian repo, I might keep the old pulseeffects package as is and treat easyeffects as a brand new software for packaging. In that case, my work would be ensuring the availability of dependencies in distros (gtkmm4 here). The packaging of easyeffects should come after that.

**Comment by Ahmed-E-86** [comment]
I tested this flatpak, and it works somehow >> https://github.com/flathub/flathub/pull/2429

There is no icon for the application, but I ran it from the terminal
`flatpak run com.github.wwmm.easyeffects`

![Screenshot from 2021-08-03 22-21-46](https://user-images.githubusercontent.com/29783425/128081376-c7a2e763-91ab-426e-9140-92bc556acd9f.png)

**Comment by vchernin** [comment]
> I tested this flatpak, and it works somehow

It is the goal that it works :)
The issue being discussed here is that building a .DEB or .RPM (or similar) using distro packages is difficult, mainly due to the recent dependencies required. To get EasyEffects building in Flatpak some manual dependency hunting was needed, however once  was done it should work without hassle for users. Unfortunately the fact that it works in Flatpak doesn't necessarily help distro packages.

> There is no icon for the application, but I ran it from the terminal

Can you see an icon in the Linux Mint software manager (probably under installed apps)? The Flatpak's icon should appear along your other icons.

> It spams a lot of warnings about missing plugins  

See https://github.com/wwmm/easyeffects/issues/1042

> I use Linux Mint 20.2 which is based on old version of Ubuntu (20.04)

Unless you have explictly installed PipeWire for audio EasyEffects won't be very useful for you, and Ubuntu 20.04's version is a bit too old. However given your app seems active I presume you've installed PipeWire through a PPA or something.

**Comment by Ahmed-E-86** [comment]
> Can you see an icon in the Linux Mint software manager (probably under installed apps)? The Flatpak's icon should appear along your other icons. Could you try installing PulseEffects Flatpak and seeing if that icon shows up on Mint?

The icon appeared after restarting my machine.

> Unless you have explictly installed PipeWire for audio EasyEffects won't be very useful for you, and Ubuntu 20.04's version is a bit too old. However given your app seems active I presume you've installed PipeWire through a PPA or something.

I installed it using this ppa >> https://launchpad.net/~pipewire-debian/+archive/ubuntu/pipewire-upstream

**Comment by v1gnesh** [comment]
Sharing a link to a related item - https://www.jelmer.uk/fresh-builds.html

**Comment by Dhipz** [comment]
I've manage to build EasyEffects after building some dependancies (glibmm, glib2, gtk4, gtkmm, lv2, rnnoise, zita-convolver) and the rest using repo but i got an error 

```
(easyeffects:5776): GLib-GIO-ERROR **: 15:43:28.580: Settings schema 'com.github.wwmm.easyeffects' is not installed
[1]    5776 trace trap (core dumped)  ./easyeffects
```

I'm not sure why since i can run `glib-compile-schemas data/schemas/` just fine. Any idea?

**Comment by wwmm** [comment]
@Dhipz did you run `source ../../util/environmental_variables.sh`? As the build folder isn't a among the system paths GSettings searches for schemas we have to tell it where the local schema is. `environmental_variables.sh` helps with that when running from the build folder.

**Comment by Dhipz** [comment]
Lol yeah, sorry I forgot to source it. Haven't tested all the plugins, but it seems like it's working perfectly except the no app icon.

![Screenshot from 2021-11-17 19-22-19](https://user-images.githubusercontent.com/45262661/142200592-c39afdac-ebb7-439f-a0dc-a5bf190d3612.png)

I'm sorry but i don't think i can help with building the packages since I basically don't know what I'm doing lol. 

If it helps, I'm using `Ubuntu 20.04 5.11.0-40-generic` and what i did was follows @wwmm's instruction and tried to installs all the newest version dependencies and its dependencies.
Some of the dependencies needed to be build from source are glibmm, glib2, gtk4, gtkmm, lv2, rnnoise, zita-convolver, and probably others that i forgot.

Good luck all!

**Comment by bhack** [comment]
> In that case, my work would be ensuring the availability of dependencies in distros (gtkmm4 here). The packaging of easyeffects should come after that.

@hosiet next release doesn't require gtkmm. Please take a look at https://github.com/wwmm/easyeffects/issues/1243

**Comment by hosiet** [comment]
Thanks for the info. I will follow the updates there.

**Comment by wwmm** [comment]
The current master branch does not use gtkmm and glibmm anymore. We now interact with gtk4 directly. This should make things easier for Debian based distributions.

**Comment by tilgovi** [comment]
I've got WIP at https://github.com/tilgovi/easyeffects/tree/issue-1000.

Open issues are the lack of libsigc++3 and rnnoise in Debian. I think we could patch out rnnoise, but I don't know what to do about libsigc++3.

**Comment by hosiet** [comment]
rnnoise is unlikely to enter Debian in the foreseeable future due to its non-free nature as stated in https://bugs.debian.org/980839 , unless things has changed since then.

libsigc++ 3.x _may_ be packaged in Debian with some work. I will take a look at it.

**Comment by hosiet** [comment]
Update: I prepared packaging for libsigc++-3.0 for Debian, which is being reviewed in the NEW queue https://ftp-master.debian.org/new.html . Once the review is complete, we will be able to find libsigc++-3.0 in future Debian and Ubuntu.

There isn't really a solution to rnnoise; I wish easyeffects could make this dependency optional in some way.

**Comment by wwmm** [comment]
>There isn't really a solution to rnnoise; I wish easyeffects could make this dependency optional in some way.

I will try to take a look at it. But I am not sure when because I have been very busy at my work. In any case this is going to require some code refactoring and the usage of compiler preprocessor directives to selectively enable only some parts of the code during compilation. And our Meson scripts will probably have to be updated for all of this as well.

In the past rnnoise was optional because it was used through a custom GStreamer plugin. And like any other GStreamer plugin it did not have to be available when compiling our code. Now that is is being used through the native PipeWire API things got more coupled to the rest of the code. The same thing happened to the convolver. The difference is that the `zita-convolver` library does not seem to have the same license problems that now I now rnnoise has.

**Comment by wwmm** [comment]
@hosiet I have updated the master branch with a code that as far as could see made rnnoise optional. When possible try to see if that is really the case. 

As usual the code readability has become worse with the preprocessor `ifdef/endif`... But it is life. I doubt something else could have been done to fix this situation once and for all...

At least for now it is still possible to add the noise reduction controls to the window even when it is disabled at compile time. Some users will probably be confused by this and may not understand why the plugin is not working. But adding this level of consistency will require much more work...

**Comment by hosiet** [comment]
So now libsigc++-3.0 is just accepted in Debian at https://tracker.debian.org/pkg/libsigc++-3.0 . I will be looking into whether easyeffects is now buildable / usable.

**Comment by hosiet** [comment]
Now easyeffects appears in Debian's official repository at https://tracker.debian.org/pkg/easyeffects . The package will be picked up by downstream deb distributions (e.g., Ubuntu 22.10+) later.

**Comment by swfsql** [comment]
It appears that the package will exist in the kinetic suite (2022-10-22) will have the package.  
https://packages.ubuntu.com/kinetic/easyeffects
